eng/common/pipelines/templates/steps/update-docsms-metadata.yml

```diff
@@ -0,0 +1,103 @@
+parameters:
+  - name: Artifacts
+    type: object
+    default: []
+  - name: RepoId
+    type: string
+    default: $(Build.Repository.Name)
+  - name: WorkingDirectory
+    type: string
+    default: ''
+  - name: ScriptDirectory
+    type: string
+    default: eng/common/scripts
+  - name: TargetDocRepoName
+    type: string
+    default: ''
+  - name: TargetDocRepoOwner
+    type: string
+  - name: Language
+    type: string
+    default: ''
+  - name: DailyDocsBuild
+    type: boolean
+    default: false
+  - name: SparseCheckoutPaths
+    type: object
+    default:
+      - '**'
+
+steps:
+- template: /eng/common/pipelines/templates/steps/enable-long-path-support.yml
+
+- pwsh: |
+    Write-Host "###vso[task.setvariable variable=DocRepoLocation]${{ parameters.WorkingDirectory }}/doc"
+  displayName: Set $(DocRepoLocation)
+
+- template: /eng/common/pipelines/templates/steps/sparse-checkout.yml
+  parameters:
+    SkipDefaultCheckout: true
+    Repositories:
+      - Name: ${{ parameters.TargetDocRepoOwner }}/${{ parameters.TargetDocRepoName }}
+        WorkingDirectory: $(DocRepoLocation)
+    Paths: ${{ parameters.SparseCheckoutPaths }}
+
+# If performing a daily docs build set the $(TargetBranchName) to a daily branch
+# name and attempt to checkout the daily docs branch. If the branch doesn't
+# exist, create it
+- ${{ if eq(parameters.DailyDocsBuild, 'true') }}:
+  - template: /eng/common/pipelines/templates/steps/set-daily-docs-branch-name.yml
+    parameters:
+      DailyBranchVariableName: TargetBranchName
+
+  - pwsh: |
+      $ErrorActionPreference = "Continue"
+      $RemoteName = "origin"
+      $BranchName = "$(TargetBranchName)"
+      # Fetch and checkout remote branch if it already exists otherwise create a new branch.
+      git ls-remote --exit-code --heads $RemoteName $BranchName
+      if ($LASTEXITCODE -eq 0) {
+        Write-Host "git fetch $RemoteName $BranchName"
+        git fetch $RemoteName $BranchName
+        Write-Host "git checkout $BranchName."
+        git checkout $BranchName
+      } else {
+        Write-Host "git checkout -b $BranchName."
+        git checkout -b $BranchName
+      }
+    displayName: Checkout daily docs branch if it exists
+    workingDirectory: $(DocRepoLocation)
+
+# If NOT performing a daily docs build, set the $(TargetBranchName) to the 
+# default branch of the documentation repository.
+- ${{ if ne(parameters.DailyDocsBuild, 'true') }}:
+  - template: /eng/common/pipelines/templates/steps/set-default-branch.yml
+    parameters:
+      WorkingDirectory: $(DocRepoLocation)
+
+  - pwsh: |
+      Write-Host "##vso[task.setvariable variable=TargetBranchName;]$(DefaultBranch)"
+    displayName: Set $(TargetBranchName) as $(DefaultBranch)
+
+- pwsh: |
+    $artifactInfoJson = '${{ convertToJson(parameters.Artifacts) }}'.Trim('"')
+    $artifactInfo = ConvertFrom-Json $artifactInfoJson
+    foreach($package in $artifactInfo) {
```

Comment:
@weshaggard -- Putting a loop in `Update-DocsMsMetadata.ps1` does work, but it requires that all usages of `Update-DocsMsMetadata.ps1` phrase the work in terms of a list. I think it makes sense to have the command perform a task on a single item and that looping logic should be handled by an external context (in this case a loop in a powershell script in DevOps, but not the kind that generates a list of DevOps steps). What do you think?

Reply:
After chatting w/ @benbp the looping (business) logic belongs inside the powershell script. Powershell will automatically handle a single object or an array of objects so usability does not suffer.